From ba34631dc188b92633de061bbe64dc6e5d893910 Mon Sep 17 00:00:00 2001
From: MoozIiSP <yuaolni@gmail.com>
Date: Mon, 27 Apr 2026 08:34:36 +0800
Subject: [PATCH 6/7] power: supply: axp20x_battery: implement calibration support

---
 drivers/power/supply/axp20x_battery.c | 38 +++++++++++++++++++++++++--
 1 file changed, 36 insertions(+), 2 deletions(-)

diff --git a/drivers/power/supply/axp20x_battery.c b/drivers/power/supply/axp20x_battery.c
index 3712a1851..aabbccddee 100644
--- a/drivers/power/supply/axp20x_battery.c
+++ b/drivers/power/supply/axp20x_battery.c
@@ -42,6 +42,10 @@
 #define AXP717_PWR_OP_BATT_PRESENT	BIT(3)
 
+#define AXP228_FULL_CAPACITY_CALIBRATE_EN	BIT(2)
+#define AXP228_CAPACITY_CALIBRATE		BIT(1)
+#define AXP228_CALIBRATE_MASK			GENMASK(2, 1)
+
 #define AXP717_BATT_PMU_FAULT_MASK	GENMASK(2, 0)
 #define AXP717_BATT_UVLO_2_5V		BIT(2)
 #define AXP717_BATT_OVER_TEMP		BIT(1)
@@ -120,6 +124,7 @@ struct axp20x_batt_ps {
 	unsigned int max_ccc;
 	const struct axp_data	*data;
+	unsigned int energy_full_design;
 	bool ts_disable;
 };
 
@@ -417,6 +422,10 @@ static int axp20x_battery_get_prop(struct power_supply *psy,
 
 		val->intval = (reg & AXP209_FG_PERCENT) * 1000; /* uAh */
 		break;
+	case POWER_SUPPLY_PROP_ENERGY_FULL_DESIGN:
+		val->intval = axp20x_batt->energy_full_design;
+		break;
+
 	case POWER_SUPPLY_PROP_VOLTAGE_MAX:
 		return axp20x_batt->data->get_max_voltage(axp20x_batt,
 							  &val->intval);
@@ -592,6 +601,10 @@ static int axp717_battery_get_prop(struct power_supply *psy,
 
 		val->intval = (reg & AXP209_FG_PERCENT) * 1000; /* uAh */
 		break;
+	case POWER_SUPPLY_PROP_ENERGY_FULL_DESIGN:
+		val->intval = axp20x_batt->energy_full_design;
+		break;
+
 	case POWER_SUPPLY_PROP_VOLTAGE_MAX:
 		return axp20x_batt->data->get_max_voltage(axp20x_batt,
 							  &val->intval);
@@ -821,6 +834,16 @@ static int axp20x_battery_set_prop(struct power_supply *psy,
 {
 	struct axp20x_batt_ps *axp20x_batt = power_supply_get_drvdata(psy);
+	unsigned int val1;
 
 	switch (psp) {
 	case POWER_SUPPLY_PROP_VOLTAGE_MIN:
@@ -830,6 +853,15 @@ static int axp20x_battery_set_prop(struct power_supply *psy,
 	case POWER_SUPPLY_PROP_CONSTANT_CHARGE_CURRENT_MAX:
 		return axp20x_set_max_constant_charge_current(axp20x_batt,
 							      val->intval);
+	case POWER_SUPPLY_PROP_CALIBRATE:
+		if (val->intval)
+			val1 = AXP228_FULL_CAPACITY_CALIBRATE_EN | AXP228_CAPACITY_CALIBRATE;
+		else
+			val1 = 0;
+		return regmap_update_bits(axp20x_batt->regmap, AXP20X_CC_CTRL,
+					  AXP228_CALIBRATE_MASK, val1);
+
 	case POWER_SUPPLY_PROP_STATUS:
 		switch (val->intval) {
 		case POWER_SUPPLY_STATUS_CHARGING:
@@ -860,6 +892,16 @@ static int axp717_battery_set_prop(struct power_supply *psy,
 {
 	struct axp20x_batt_ps *axp20x_batt = power_supply_get_drvdata(psy);
+	unsigned int val1;
 
 	switch (psp) {
 	case POWER_SUPPLY_PROP_VOLTAGE_MIN:
@@ -871,6 +913,15 @@ static int axp717_battery_set_prop(struct power_supply *psy,
 	case POWER_SUPPLY_PROP_CONSTANT_CHARGE_CURRENT_MAX:
 		return axp717_set_constant_charge_current(axp20x_batt,
 							  val->intval);
+	case POWER_SUPPLY_PROP_CALIBRATE:
+		if (val->intval)
+			val1 = AXP228_FULL_CAPACITY_CALIBRATE_EN | AXP228_CAPACITY_CALIBRATE;
+		else
+			val1 = 0;
+		return regmap_update_bits(axp20x_batt->regmap, AXP20X_CC_CTRL,
+					  AXP228_CALIBRATE_MASK, val1);
+
 	case POWER_SUPPLY_PROP_STATUS:
 		switch (val->intval) {
 		case POWER_SUPPLY_STATUS_CHARGING:
@@ -915,7 +966,8 @@ static int axp20x_battery_prop_writeable(struct power_supply *psy,
 	       psp == POWER_SUPPLY_PROP_VOLTAGE_MIN ||
 	       psp == POWER_SUPPLY_PROP_VOLTAGE_MAX ||
 	       psp == POWER_SUPPLY_PROP_CONSTANT_CHARGE_CURRENT ||
-	       psp == POWER_SUPPLY_PROP_CONSTANT_CHARGE_CURRENT_MAX;
+	       psp == POWER_SUPPLY_PROP_CONSTANT_CHARGE_CURRENT_MAX ||
+	       psp == POWER_SUPPLY_PROP_CALIBRATE;
 }
 
 static int axp717_battery_prop_writeable(struct power_supply *psy,
@@ -924,7 +976,8 @@ static int axp717_battery_prop_writeable(struct power_supply *psy,
 	return psp == POWER_SUPPLY_PROP_STATUS ||
 	       psp == POWER_SUPPLY_PROP_VOLTAGE_MIN ||
 	       psp == POWER_SUPPLY_PROP_VOLTAGE_MAX ||
-	       psp == POWER_SUPPLY_PROP_CONSTANT_CHARGE_CURRENT_MAX;
+	       psp == POWER_SUPPLY_PROP_CONSTANT_CHARGE_CURRENT_MAX ||
+	       psp == POWER_SUPPLY_PROP_CALIBRATE;
 }
 
 static const struct power_supply_desc axp209_batt_ps_desc = {
-- 
2.50.1 (Apple Git-155)
